Create Basic Data Types Tutorial (#1514)
ocaml/ocaml.org: data/tutorials/language/0lg_00_basic_datatypes.md

There are no dedicated types for unsigned integers in OCaml. Bitwise operations on `int` treat the sign bit the same as other bits. Binary operators use standard symbols. The signed remainder operator is written `mod`. Integers in OCaml have no predefined power operator.

#### Floats and Type Conversions

Float numbers have type `float`.

OCaml does not perform any implicit type conversion between values. Therefore, arithmetic expressions can't mix integers and floats. Parameters are either all `int` or all `float`. Arithmetic operators on floats are not the same, and they are written with a dot suffix: `+.`,  `-.`, `*.`, `/.`.
```ocaml
# let pi = 3.14159;;
val pi : float = 3.14159

# let tau = 2.0 *. pi;;
val tau : float = 6.28318

# let tau = 2 *. pi;;
Error: This expression has type int but an expression was expected of type
         float

# let tau = 2 * pi;;
Error: This expression has type float but an expression was expected of type
         int
```

> Operations on `float` are provided by the [`Stdlib`](/api/Stdlib.html) and the [`Float`](/api/Float.html) modules.

Reviewer: This sentences is repeated from li.49. Which is the better place for it?

```ocaml
# type character_class =
    | Barbarian
    | Bard
    | Cleric
    | Druid
    | Fighter
    | Monk
    | Paladin
    | Ranger
    | Rogue
    | Sorcerer
    | Wizard;;
type character_class =
    Barbarian
  | Bard
  | Cleric
  | Druid
  | Fighter
  | Monk
  | Paladin
  | Ranger
  | Rogue
  | Sorcerer
  | Wizard

# type rectitude = Evil | R_Neutral | Good;;
type rectitude = Evil | R_Neutral | Good

# type firmness = Chaotic | F_Neutral | Lawful;;
type firmness = Chaotic | F_Neutral | Lawful
```

These kinds of variant types can also be used to represent weekdays, cardinal
directions, or any other fixed-sized set of values that can be given names. An
ordering is defined on values following the definition order (e.g., `Druid
< Ranger`).

Here is how pattern matching can be performed on the types defined above.
```ocaml
# let rectitude_to_french = function
    | Evil -> "Mauvais"
    | R_Neutral -> "Neutre"
    | Good -> "Bon";;
val rectitude_to_french : rectitude -> string = <fun>
```

Note that:

- `unit` is a variant with a unique constructor, which does not carry data: `()`.
- `bool` is also a variant with two constructors that don't carry data: `true` and `false`.

> A pair `(x, y)` has type `a * b`, where `a` is the type of `x` and `b` is the type of `y`. Some may find it intriguing that `a * b` is called a product. Although this is not a complete explanation, here is a remark that may help in understanding: Consider the product type `character_class * character_alignement`. There are 12 classes and 9 alignments. Any pair of values from those types inhabits the product type. Therefore, in the product type, there are 9 × 12 = 108 values, which is also a product.

Reviewer: ```suggestion
A pair `(x, y)` has type `a * b`, where `a` is the type of `x` and `b` is the type of `y`. Some may find it intriguing that `a * b` is called a product. Although this is not a complete explanation, this may help in understanding: Consider the product type `character_class * character_alignement`. There are 12 classes and 9 alignments. Any pair of values from those types inhabits the product type. Therefore, in the product type, there are 9 × 12 = 108 values, which is also a product.
```

1. Records **should not** be defined using this technique. It is only demonstrated here to further illustrate the expressive strength of OCaml variants.

1. This way, to define records **may** be applied to _Generalised Algebraic Data Types_, which is the subject of another tutorial.

### Type Aliases

Just like values, any type can be given a name.
```ocaml
# type latitude_longitude = float * float;;
type latitude_longitude = float * float
```

This is mostly useful as a means of documentation or as a means to shorten long-type expressions.

## A Complete Example: Mathematical Expressions

This example show how to represent simple mathematical expressions like `n * (x + y)` and multiply them out symbolically to get `n * x + n * y`.

Here is a type for these expressions:
```ocaml env=expr
# type expr =
  | Plus of expr * expr        (* a + b *)
  | Minus of expr * expr       (* a - b *)
  | Times of expr * expr       (* a * b *)
  | Divide of expr * expr      (* a / b *)

>   | Var of string              (* "x", "y", etc. *);;

Reviewer: ```suggestion
  | Var of string              (* "x", "y", etc. *);;
type expr =
    Plus of expr * expr
  | Minus of expr * expr
  | Times of expr * expr
  | Divide of expr * expr
  | Var of string
```

  | Minus (e1, e2) -> "(" ^ to_string e1 ^ " - " ^ to_string e2 ^ ")"
  | Times (e1, e2) -> "(" ^ to_string e1 ^ " * " ^ to_string e2 ^ ")"
  | Divide (e1, e2) -> "(" ^ to_string e1 ^ " / " ^ to_string e2 ^ ")"
  | Var v -> v;;
val to_string : expr -> string = <fun>
```

We can write a function to multiply out expressions of the form `n * (x + y)`
or `(x + y) * n`, and for this we will use a nested pattern:

```ocaml env=expr
# let rec distribute = function
  | Times (e1, Plus (e2, e3)) ->
     Plus (Times (distribute e1, distribute e2),
           Times (distribute e1, distribute e3))
  | Times (Plus (e1, e2), e3) ->
     Plus (Times (distribute e1, distribute e3),
           Times (distribute e2, distribute e3))
  | Plus (e1, e2) -> Plus (distribute e1, distribute e2)
  | Minus (e1, e2) -> Minus (distribute e1, distribute e2)
  | Times (e1, e2) -> Times (distribute e1, distribute e2)
  | Divide (e1, e2) -> Divide (distribute e1, distribute e2)
  | Var v -> Var v;;
val distribute : expr -> expr = <fun>
```

This is how it can be used:
```ocaml env=expr
# e |> distribute |> to_string |> print_endline;;
((n * x) + (n * y))
- : unit = ()
```

> How does the `distribute` function work? The key is in the first two patterns.

Reviewer: ```suggestion
The first two patterns hold the key to how the `distribute` function works.
```

From the data point of view, records and tuples are similar to the logical conjunction “and,” while variants are similar to the logical disjunction “or.” This analogy goes very deep, with records and tuples on one side as products and variants on the other side as union. These are true mathematical operations on data types. Records and tuples play the role of multiplication, which is why they are called product types. Variants play the role of addition. Putting it all together, basic OCaml types are said to be algebraic.

## Next: Advanced Data Types

Going further, there are several advanced topics related to data types in OCaml that you can explore to deepen your understanding and enhance your programming skills.
- Mutually Recursive Variants
- Polymorphic Variants
- Extensible Variants
- Generalised Algebraic Data Types

> As of writing this, these topics are not yet covered in tutorials. Documentation on them is available in the OCaml [Language Manual](https://v2.ocaml.org/releases/5.0/htmlman/index.html) and in the [books](https://ocaml.org/books) on OCaml.

Reviewer: ```suggestion
As of writing this, these topics will be covered in forthcoming tutorials. Documentation on them is available in the OCaml [Language Manual](https://v2.ocaml.org/releases/5.0/htmlman/index.html) and in the [books](https://ocaml.org/books) on OCaml.
```
Do you like this better?
